pkg/controller/machine_safety_test.go

```diff
@@ -71,6 +71,121 @@ var _ = Describe("machine", func() {
 		Entry("one machineset", nil, map[string]string{"machineset-1": ""}, nil, []string{"machineset-1"}, nil),
 	)
 
+	DescribeTable("##unfreezeMachineSetsAndDeployments",
+		func(machineSetExists, machineSetIsFrozen, parentExists, parentIsFrozen bool) {
+			createController := func(stop <-chan struct{}, objects ...runtime.Object) *controller {
+				fakeUntypedClient := fakeuntyped.NewSimpleClientset(objects...)
+				fakeTypedClient := &faketyped.FakeMachineV1alpha1{
+					&fakeUntypedClient.Fake,
+				}
+
+				controlMachineInformerFactory := machineinformers.NewSharedInformerFactory(fakeUntypedClient, 100*time.Millisecond)
+				defer controlMachineInformerFactory.Start(stop)
+
+				machineSharedInformers := controlMachineInformerFactory.Machine().V1alpha1()
+				machineSets := machineSharedInformers.MachineSets()
+				machineDeployments := machineSharedInformers.MachineDeployments()
+
+				return &controller{
+					controlMachineClient:    fakeTypedClient,
+					machineSetLister:        machineSets.Lister(),
+					machineDeploymentLister: machineDeployments.Lister(),
+					machineSetSynced:        machineSets.Informer().HasSynced,
+					machineDeploymentSynced: machineDeployments.Informer().HasSynced,
+				}
+			}
+
+			testMachineSet := &machinev1.MachineSet{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "testMachineSet",
+					Namespace: machinenamespace,
+					Labels: map[string]string{
+						"name": "testMachineDeployment",
+					},
+				},
+				Spec: machinev1.MachineSetSpec{
+					Selector: &metav1.LabelSelector{
+						MatchLabels: map[string]string{
+							"name": "testMachineDeployment",
+						},
+					},
+				},
+			}
+			if machineSetIsFrozen {
+				testMachineSet.Labels["freeze"] = "True"
+				msStatus := testMachineSet.Status
+				mscond := NewMachineSetCondition(machinev1.MachineSetFrozen, machinev1.ConditionTrue, "testing", "freezing the machineset")
+				SetCondition(&msStatus, mscond)
+			}
+
+			testMachineDeployment := &machinev1.MachineDeployment{
```

Comment:
It might be better to add a newMachineDeployment function like the existing newMachineSet and integrate it with newIntialObjects.

I will do this and you can maybe rebase after that.

Reply:
same as above